### PR TITLE
fix: generate xml report packages correctly on windows

### LINKED ISSUE #1573: Duplicate package elements for same name in XML report on Windows
When generating an xml report on Windows, the package element for a name is generated twice:

```<?xml version="1.0" ?>
<coverage version="7.2.1" timestamp="1677844898119" lines-valid="848" lines-covered="807" line-rate="0.9517" branches-valid="201" branches-covered="166" branch-rate="0.8259" complexity="0">
	<!-- Generated by coverage.py: https://coverage.readthedocs.io/en/7.2.1 -->
	<!-- Based on https://raw.githubusercontent.com/cobertura/web/master/htdocs/xml/coverage-04.dtd -->
	<sources>
		<source>(...)</source>
	</sources>
	<packages>
		<package name="cd.python" line-rate="0.9511" branch-rate="0.8259" complexity="0">
			<classes>
			(...)
			</classes>
		</package>
		<package name="cd.python.interpreter" line-rate="1" branch-rate="1" complexity="0">
			<classes>
			(...)
			</classes>
		</package>
		<package name="cd.python" line-rate="1" branch-rate="1" complexity="0">
			<classes>
			(...)
			</classes>
		</package>
	</packages>
</coverage>
```

This is generating a problem, when the report needs to be read with the Jenskins Code Coverage API plugin, as only the last element with the same name is used. The last element usually only consists of one file.

On debugging the `XMLReporter` it seems one package (with one file) is generated with a wrong package name:
coveragepy_xml_duplicate_package

This seems to be due to the `fr.relative_filename()` missing the path name conversion.

I'll provide a pull request for this issue.

**To Reproduce**
As the code I am working on is proprietary, I have no code base to test this, but I hope my description was good enough.

Python version: Python 3.10.10 (tags/v3.10.10:aad5f6a, Feb  7 2023, 17:20:36) [MSC v.1929 64 bit (AMD64)] on win32
Coverage version: 7.2.1

**Expected behavior**
Expected behavior is only one package element per individual package:

```<?xml version="1.0" ?>
<coverage version="7.2.1" timestamp="1677844898119" lines-valid="848" lines-covered="807" line-rate="0.9517" branches-valid="201" branches-covered="166" branch-rate="0.8259" complexity="0">
	<!-- Generated by coverage.py: https://coverage.readthedocs.io/en/7.2.1 -->
	<!-- Based on https://raw.githubusercontent.com/cobertura/web/master/htdocs/xml/coverage-04.dtd -->
	<sources>
		<source>(...)</source>
	</sources>
	<packages>
		<package name="cd.python" line-rate="0.9511" branch-rate="0.8259" complexity="0">
			<classes>
			(...)
			</classes>
		</package>
		<package name="cd.python.interpreter" line-rate="1" branch-rate="1" complexity="0">
			<classes>
			(...)
			</classes>
		</package>
	</packages>
</coverage>
```

### PR DESCRIPTION
**Pull request by bparzella**
This will fix the problem of duplicate package elements with the same name in the xml report generated on windows.

Closes #1573

### CODE REVIEW

File: coverage/xmlreport.py [hunk 1]
```diff
@@ -182,7 +182,7 @@ def xml_file(self, fr: FileReporter, analysis: Analysis, has_arcs: bool) -> None
                 rel_name = filename[len(source_path)+1:]
                 break
         else:
-            rel_name = fr.relative_filename()
```
Comment (bparzella) [inline coverage/xmlreport.py:185]: If the for loop is run, the `rel_name` is generated from `filename` which is already converted from backslash to slash in line 177

When the else branch is run, `rel_name` is generated from the `FileReporter` which needs the conversion

**Comment by blakeNaccarato** [comment]
Coming here from [the Coverage.py Mastodon request](https://hachyderm.io/@coveragepy/109999060333074351) for someone on a Windows machine to validate this. I cloned this repo and checked out the PR. I installed it in a virtual environment on Python 3.11 with `pip install .` and ran `pytest`. It works fine, with the result `1299 passed, 28 skipped in 70.21s (0:01:10)`.

@nedbat Is there any log file I should share that would help satisfy the request to validate the fix? Or are you looking for it to be tested in an actual use-case scenario, rather than in the test suite?

Also, click to expand the details below on issues I had even getting the dev installation to work on my Windows machine, which is why I fell back to good old `pip install .`. Neither installing `requirements/dev.pip` nor running `tox` worked on my machine, due to `pywin32-ctypes>=0.2.0` not being `==` pinned. A separate issue entirely, but I just thought I'd point it out.

<details><summary>Notes on dev install not working...</summary>

I followed the setup instructions

```PowerShell
py -3.11 -m venv .venv
.\.venv\Scripts\activate
pip install -r requirements/dev.pip
```

But I got an error from `pip install -r requirements/dev.pip` about `pywin32-ctypes>=0.2.0` not being `==` pinned.

```
ERROR: In --require-hashes mode, all requirements must have their versions pinned with ==. These do not:
    pywin32-ctypes>=0.2.0 from https://files.pythonhosted.org/packages/9e/4b/3ab2720f1fa4b4bc924ef1932b842edf10007e4547ea8157b0b9fc78599a/pywin32_ctypes-0.2.0-py2.py3-none-any.whl (from keyring==23.13.1->-r requirements/dev.pip 
(line 292))
```

So I ran `pip install -r requirements/dev.pip` instead, but got essentially the same error after running `tox`

```
ERROR: In --require-hashes mode, all requirements must have their versions pinned with ==. These do not:
    pywin32-ctypes>=0.2.0 from https://files.pythonhosted.org/packages/9e/4b/3ab2720f1fa4b4bc924ef1932b842edf10007e4547ea8157b0b9fc78599a/pywin32_ctypes-0.2.0-py2.py3-none-any.whl (from keyring==23.13.1->-r requirements/lint.pip (line 330))
lint: exit 1 (35.52 seconds) C:\Users\<...>\Code\other\coveragepy> python -m pip install -U -r requirements/lint.pip pid=12740
.pkg: _exit> python C:\Users\<...>\Code\other\coveragepy\.venv\Lib\site-packages\pyproject_api\_backend.py True setuptools.build_meta
  py37: SKIP (0.02 seconds)
  py38: SKIP (0.02 seconds)
  py39: SKIP (0.05 seconds)
  py310: SKIP (0.00 seconds)
  py311: OK (203.17=setup[63.98]+cmd[0.61,4.41,14.73,70.73,0.39,48.31] seconds)
  py312: SKIP (0.02 seconds)
  pypy3: SKIP (0.00 seconds)
  doc: FAIL code 2 (58.92=setup[45.66]+cmd[0.48,12.78,0.00] seconds)
  lint: FAIL code 1 (36.34 seconds)
  evaluation failed :( (299.25 seconds)
```

So I just installed `coveragepy` directly into my `.venv` with `pip install .`. I also ran `pip install -r requirements/pytest.pip`. Upon running `pytest`

</details>

**Comment by nedbat** [comment]
@blakeNaccarato thanks!  Did you try to reproduce the error from the original issue #1573?  We have CI that confirms the pull request works, but it's hard to demonstrate the problem first.

@bparzella Maybe we could use your zip file from #1573 as the basis for a new test?

**Comment by blakeNaccarato** [comment]
I can verify on my Windows machine that `<package name="namespace.package1"...>` is duplicated when installing `pip install coverage`, but not when installing `pip install git+https://github.com/bparzella/coveragepy.git@xml_duplicate_fix`. See [coverage_test_compare.zip](https://github.com/nedbat/coveragepy/files/10944801/coverage_test_compare.zip) where I wrote a script to run both versions in separate virtual environments and diff'ed the resulting `coverage.xml`. The diff shows an unequal occurrence of `<package name="namespace.package1"...>` between the two runs.

A more sophisticated way to test this would be to check the resulting `xml` for duplicate entries rather than a full diff.

<details><summary>.diff</summary>

```DIFF
diff --git a/pypi/proj/coverage.xml b/xml_duplicate_fix/proj/coverage.xml
index 5827aa2..c7f43c6 100644
--- a/pypi/proj/coverage.xml
+++ b/xml_duplicate_fix/proj/coverage.xml
@@ -1,6 +1,6 @@
...
@@ -8,32 +8,34 @@
 	<packages>
 		<package name="namespace.package1" line-rate="1" branch-rate="0" complexity="0">
 			<classes>
-				<class name="__main__.py" filename="namespace/package1/__main__.py" complexity="0" line-rate="1" branch-rate="0">
+				<class name="__init__.py" filename="namespace/package1/__init__.py" complexity="0" line-rate="1" branch-rate="0">
 					<methods/>
 					<lines>
...
 					</lines>
 				</class>
-				<class name="sample.py" filename="namespace/package1/sample.py" complexity="0" line-rate="1" branch-rate="0">
+				<class name="__main__.py" filename="namespace/package1/__main__.py" complexity="0" line-rate="1" branch-rate="0">
 					<methods/>
 					<lines>
...
 					</lines>
 				</class>
-				<class name="test.py" filename="namespace/package1/test.py" complexity="0" line-rate="1" branch-rate="0">
+				<class name="sample.py" filename="namespace/package1/sample.py" complexity="0" line-rate="1" branch-rate="0">
 					<methods/>
 					<lines>
...
 					</lines>
 				</class>
-			</classes>
-		</package>
-		<package name="namespace.package1.subpackage" line-rate="1" branch-rate="0" complexity="0">
-			<classes>
-				<class name="test.py" filename="namespace/package1/subpackage/test.py" complexity="0" line-rate="1" branch-rate="0">
+				<class name="test.py" filename="namespace/package1/test.py" complexity="0" line-rate="1" branch-rate="0">
 					<methods/>
 					<lines>
 						<line number="1" hits="1"/>
@@ -42,19 +44,13 @@
 				</class>
 			</classes>
 		</package>
-		<package name="namespace.package1" line-rate="1" branch-rate="0" complexity="0">
+		<package name="namespace.package1.subpackage" line-rate="1" branch-rate="0" complexity="0">
 			<classes>
-				<class name="__init__.py" filename="namespace/package1/__init__.py" complexity="0" line-rate="1" branch-rate="0">
+				<class name="test.py" filename="namespace/package1/subpackage/test.py" complexity="0" line-rate="1" branch-rate="0">
 					<methods/>
 					<lines>
...
 					</lines>
 				</class>
 			</classes>

```

</details>

**Comment by bparzella** [comment]
> @bparzella Maybe we could use your zip file from #1573 as the basis for a new test?

@nedbat That sounds like a good idea, I'll look into it.

**Comment by bparzella** [comment]
Running the test on windows using the master branch has the following results:

```
py310: commands[3]> python igor.py test_with_tracer c tests\test_xml.py::XmlReportTest::test_no_duplicate_packages
=== CPython 3.10.10 (rev aad5f6a) with C tracer (.tox\py310\Scripts\python.EXE) ===
bringing up nodes...
bringing up nodes...

F                                                                                                                                                                                                                                       [100%]
================================================================================================================== FAILURES ================================================================================================================== 
__________________________________________________________________________________________________ XmlReportTest.test_no_duplicate_packages __________________________________________________________________________________________________ 
[gw0] win32 -- Python 3.10.10 (...)\.tox\py310\Scripts\python.EXE

self = <tests.test_xml.XmlReportTest object at 0x000002C6ECD8A290>

    def test_no_duplicate_packages(self) -> None:
        self.make_file(
            "namespace/package/__init__.py",
            "from . import sample; from . import test; from .subpackage import test"
        )
        self.make_file("namespace/package/sample.py", "print('package.sample')")
        self.make_file("namespace/package/test.py", "print('package.test')")
        self.make_file("namespace/package/subpackage/test.py", "print('package.subpackage.test')")

        # no source path passed to coverage!
        # problem occurs when they are dynamically generated during xml report
        cov = coverage.Coverage()

        cov.start()
        import_local_file("foo", "namespace/package/__init__.py")               # pragma: nested
        cov.stop()                                                              # pragma: nested

        cov.xml_report()

        dom = ElementTree.parse("coverage.xml")

        # only two packages should be present
        packages = dom.findall(".//package")
>       assert len(packages) == 2
E       AssertionError: assert 3 == 2
E        +  where 3 = len([<Element 'package' at 0x000002C6ECD4FC40>, <Element 'package' at 0x000002C6ECEBEC00>, <Element 'package' at 0x000002C6ECEBF560>])

(...)\tests\test_xml.py:344: AssertionError
------------------------------------------------------------------------------------------------------------ Captured stdout call ------------------------------------------------------------------------------------------------------------ 
package.sample
package.test
package.subpackage.test
========================================================================================================== short test summary info =========================================================================================================== 
FAILED tests/test_xml.py::XmlReportTest::test_no_duplicate_packages - AssertionError: assert 3 == 2
```

**Comment by nedbat** [comment]
This is now released as part of [coverage 7.2.2](https://pypi.org/project/coverage/7.2.2).
